libnetwork/drivers/bridge/port_mapping_linux_test.go

```diff
@@ -171,3 +178,567 @@ func loopbackUp() error {
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 80, HostIP: net.IPv4zero},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 80, HostIP: net.IPv6zero, HostPort: firstEphemPort + 1},
+			},
+		},
+		{
+			name:     "disable nat",
+			epAddrV4: ctrIP4,
+			epAddrV6: ctrIP6,
+			cfg: []types.PortBinding{
+				{Proto: types.TCP, Port: 22},
+				{Proto: types.TCP, Port: 80},
+			},
+			gwMode4: gwModeRouted,
+			gwMode6: gwModeRouted,
+			expPBs: []types.PortBinding{
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 22, HostIP: net.IPv4zero},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 22, HostIP: net.IPv6zero},
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 80, HostIP: net.IPv4zero},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 80, HostIP: net.IPv6zero},
+			},
+		},
+		{
+			name:     "same ports for matching mappings with different host addresses",
+			epAddrV4: ctrIP4,
+			epAddrV6: ctrIP6,
+			cfg: []types.PortBinding{
+				// These two should both get the same host port.
+				{Proto: types.TCP, Port: 80, HostIP: newIPNet(t, "fd0c:9167:5b11::2/64").IP},
+				{Proto: types.TCP, Port: 80, HostIP: newIPNet(t, "192.168.1.2/24").IP},
+				// These three should all get the same host port.
+				{Proto: types.TCP, Port: 22, HostIP: newIPNet(t, "fd0c:9167:5b11::2/64").IP},
+				{Proto: types.TCP, Port: 22, HostIP: newIPNet(t, "fd0c:9167:5b11::3/64").IP},
+				{Proto: types.TCP, Port: 22, HostIP: newIPNet(t, "192.168.1.2/24").IP},
+				// These two should get different host ports, and the exact-port should be allocated
+				// before the range.
+				{Proto: types.TCP, Port: 12345, HostPort: 12345, HostPortEnd: 12346},
+				{Proto: types.TCP, Port: 12345, HostPort: 12345},
+			},
+			expPBs: []types.PortBinding{
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 12345, HostIP: net.IPv4zero, HostPort: 12345},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 12345, HostIP: net.IPv6zero, HostPort: 12345},
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 22, HostIP: newIPNet(t, "192.168.1.2/24").IP, HostPort: firstEphemPort},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 22, HostIP: newIPNet(t, "fd0c:9167:5b11::2/64").IP, HostPort: firstEphemPort},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 22, HostIP: newIPNet(t, "fd0c:9167:5b11::3/64").IP, HostPort: firstEphemPort},
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 80, HostIP: newIPNet(t, "192.168.1.2/24").IP, HostPort: firstEphemPort + 1},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 80, HostIP: newIPNet(t, "fd0c:9167:5b11::2/64").IP, HostPort: firstEphemPort + 1},
+				{Proto: types.TCP, IP: ctrIP4.IP, Port: 12345, HostIP: net.IPv4zero, HostPort: 12346},
+				{Proto: types.TCP, IP: ctrIP6.IP, Port: 12345, HostIP: net.IPv6zero, HostPort: 12346},
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer netnsutils.SetupTestOSContext(t)()
+
+			// Mock the startProxy function used by the code under test.
+			startProxy = func(proto string,
```

Comment:
Instead of relying on the returned port mappings to hold references to the mock proxies, keep track of `StartProxy` calls in the test itself, e.g. by tracking the calls in a map keyed by the tuple of arguments and using [`is.DeepEqual`](https://pkg.go.dev/gotest.tools/v3@v3.5.1/assert/cmp#DeepEqual) to assert that all the expected `StartProxy` calls were made and no unexpected calls were included.
```go
type proxyCall struct { proto, host, container, proxyPath string }
proxies := map[proxyCall]*atomic.Bool // proxy -> is not stopped
startProxy = func(...) (func() error, error) {
	c := proxyCall{proto, fmt.Sprintf("%v:%v", hostIP, hostPort), fmt.Sprintf("%v:%v", containerIP, containerPort), proxyPath}
	if _, ok := proxies[c]; ok { return nil, fmt.Errorf("duplicate proxy: %#v", c) }
	proxies[c] = true
	return func() error {
		if tc.expReleaseErr != "" {
			return fmt.Errorf("can't stop now")
		}
		if !proxies[c] {
			return fmt.Errrof("already stopped")
		}
		proxies[c] = false
		return nil
	}
}

Reply:
Nice! Done.